### PR TITLE
[TRTLLM-16466][fix] Profile cached KV chunks for dense MLA

### PR DESCRIPTION
**Pull request by zhaoyangwang-nvidia**
## Dev Engineer Review

- Adds chunk-aware MLA KV-cache profiling for bounded chunked prefill on SM100+.
- Removes the workspace reserve and admission cap only after profiling succeeds.
- Retains protections for unsupported or unprofiled paths, including DSA fallback, SM90 full-gather dispatch, and applicable Hopper configurations.
- Extends `get_mla_context_workspace_kv_len_cap(...)` with profiling controls.
- Review severity counts are unavailable.
- GPU integration, full CPU testing, accuracy investigation, and original-environment revalidation remain pending.

## QA Engineer Review

- Adds 24 parameterized KV-cache estimation tests for chunk sizing, context limits, workspace state, cache limits, hardware gating, and fallback behavior.
- Enables chunked prefill in `TestKimiK2.test_nvfp4`.
- No test-list files changed. Test-list registration is not confirmed.
- Coverage verdict: **needs follow-up**.

## Per-File QA Perspective

- `tensorrt_llm/_torch/pyexecutor/_util.py`: Verify the new profiling controls, chunked-prefill sizing, reserve removal, admission-cap behavior, and fallback paths.
- `tests/integration/defs/accuracy/test_llm_api_pytorch.py`: Verify Kimi K2 NVFP4 accuracy with chunked prefill enabled. Test-list registration is not confirmed.
- `tests/unittest/_torch/executor/kv_cache/test_kv_cache_estimation.py`: Verify parameterized coverage for hardware, chunking, workspace state, cache limits, MLA modes, and scheduler rounding. Test-list registration is not confirmed.

## Description

Addresses [TRTLLM-16466](https://jirasw.nvidia.com/browse/TRTLLM-16466).

Fresh-prompt KV-cache profiling does not exercise the cached-KV staging buffers used by dense MLA with prefix reuse. Enable chunked prefill in `TestKimiK2::test_nvfp4` and profile real cached prefixes before sizing the KV pool from the allocation peak.

For dense TRTLLM MLA on the SM100+ cached-KV chunk path, the profiler submits a long request per attention-DP rank. With KV chunk capacity C and query budget Q, its length is ceil(2*C/Q)*Q + Q. This exercises two full cached-KV chunks in one forward with a full query, including the overlap while tensors from the previous chunk remain live during the next expansion. With max_num_tokens=8192 and C=32768, the request has 73728 tokens and reaches 65536 cached + 8192 new tokens. Temporary KV capacity is calculated from these requests using the existing estimator. Configurations whose context limit cannot fit this request retain the applicable reserve/cap protection. GPU peak-memory comparison of this revised length remains pending.

After this profiling run completes, skip the old additional workspace reserve and attended-KV admission cap for the covered path. Preserve the C++ workspace-sizing fix. Beam search, speculative decoding, and skip-softmax use the same dense context path and are eligible; existing temporary-cache sizing already accounts for beam width and speculative token extras. Existing runtime compatibility checks still apply.

On SM100+, unprofiled chunked paths retain reserve/cap protection where the backend declares a nonzero workspace cost. DSA dense fallback cannot be covered generically by one long prompt because it may switch to sparse before reaching a full chunk. Hopper preserves its pre-PR backend-declared reserve/cap policy: dense MLA with scheduler chunking enabled does not gain a new reservation or admission cap. SM90 remains excluded from chunk-aware profiling because its cached-KV path still uses full-gather. Existing reservations for non-chunked reuse and backend-declared unbounded workspace remain unchanged on Hopper; addressing its full-gather memory risk is outside this PR. This PR does not globally delete the remaining protection.

## Test Coverage

- Added 24 parameterized cases in `tests/unittest/_torch/executor/kv_cache/test_kv_cache_estimation.py`: ADP/non-ADP request construction, context-length limits, runtime chunk dimensions, reserve gating, beam/spec/skip-softmax eligibility, hardware dispatch, and unsupported-path fallback. The existing `l0_cpu` executor-directory entry includes this file.
- Executed those cases against the changed source methods in a release container using real torch and executor request bindings: 24 passed. Methods were isolated from unrelated current-main imports; this is not a full repository pytest run. A direct current-source test-file attempt was blocked at collection because the rc25 KV-cache-manager-v2 binding lacks the `poison_reason` API required by current Python source.
- Repository pre-commit checks and `git diff --check`.
- Additional one-B300 compatibility matrix using small DeepSeek-shaped models, query budget 128, KV chunk 512, and profiling prompt 672. Dense, beam=2, MTP (both variants), NGram, SA, EAGLE3, draft-target, and skip-softmax all completed initialization and generation while exercising both full and multiple cached-KV chunks during profiling.
- DSA with/without dense fallback, short contexts, disabled chunking, and FlashInfer BF16 also completed execution. These successes do not all prove the dense-chunk profiling bound: with a DSA fallback threshold of 384, profiling reached only 256 cached tokens in dense mode before switching to absorption, but the reuse workload later reached **768 cached + 120 new** in dense mode. Keep its protection.
- Two matrix configurations failed for image/backend reasons: FlashInfer FP8 on SM103 rejects the selected fa2 MLA backend, and the old image's VANILLA MLA rejects its metadata type. These are not claims about every current-main configuration of those backends.

The compatibility matrix used the rc24 runtime with the changed request-construction methods transplanted and eligibility exclusions temporarily removed to inspect each path. EAGLE3 used explicit capture layers and non-Lite Q-LoRA. Draft-target used a dense-FFN draft with normal loading from small random-weight HF checkpoints (the existing dummy loader assumes a draft weight-sharing method), plus current-main's existing optional engine weight-loader guard. This is execution/coverage evidence, not a full current-main integration or accuracy run. GPU scripts and detailed logs are retained in the experiment workspace.

- Prior prototype experiment on 4xB300, Kimi-K2-Thinking-NVFP4, TP4 + attention DP, FP8 KV, `max_num_tokens=8192`, `max_batch_size=16`, KV fraction 0.8:

| Variant | Allocated KV pool / GPU | MMLU | GSM8K |
|---|---:|---:|---:|
| Existing reserve/cap | 49.93 GiB | 88.47% | 94.12% |
| Reserve/cap removed | 79.04 GiB | 88.47% | 93.52% |
| Chunking + chunk-aware profiling | 76.90 GiB | 88.23% | 93.40% |

The prototype recovered approximately **26.97 GiB of KV pool capacity per GPU** (107.88 GiB across four GPUs); rank-0 profiling peak increased by 2.68 GiB. All runs completed without OOM. Full-chunk and multi-chunk coverage was observed on all ranks. These measurements are from the earlier rc24 prototype, not an end-to-end run of this PR's current-main commit.

## Draft follow-up

- Run the updated Kimi-K2 integration test and full CPU tests on the final source revision.
- Investigate score/output differences before claiming no accuracy regression.
- Revalidate on the original OOM environment; the larger-memory B300 experiment did not reproduce the original OOM.
- Extend profiling coverage before removing the remaining reserve/cap infrastructure globally.

## PR Checklist

- [x] Scope, implementation, validation, and known limitations are described above.
- [x] Regression tests are included in an existing CPU CI test directory.
- [x] No public API or dependency changes.
- Final-source GPU integration and accuracy validation remain pending; see Draft follow-up above.

- [x] Please check this after reviewing the above items as appropriate for this PR.

### CODE REVIEW

**Comment by nvpohanh** [comment]
[by Codex] @eopXD Friendly reminder: could you review this PR? Thanks!

File: tensorrt_llm/_torch/pyexecutor/_util.py [hunk 1]
```diff
@@ -969,6 +974,42 @@ def _cal_max_memory(self, peak_memory, total_gpu_memory, fraction,
...
+            return None
+        # Beam search and speculative decoding use this same dense context
+        # path. Their extra KV capacity is accounted for by
+        # _get_token_num_for_estimation; neither needs a separate exclusion.
+        features = self._model_engine.attn_runtime_features
+        if not features.chunked_prefill or features.chunk_size <= 0:
+            return None
+        # MLA.forward_context uses full-gather on Hopper even when scheduler
+        # chunking is enabled. Only the SM100+ path bounds cached-KV staging.
+        if get_sm_version() < 100:
```
Comment (BowenFu) [inline tensorrt_llm/_torch/pyexecutor/_util.py:1050]: Good that the Hopper case is pinned by a test rather than assumed — `test_mla_profile_requires_bounded_runtime_path` covering `sm=90` makes the intent unambiguous.

One consequence worth calling out in the description: `TrtllmAttention.runtime_workspace_is_chunked_prefill_bounded` returns `True` for dense MLA on every SM today, so a DeepSeek run on H100/H200 with `enable_chunked_prefill=True` currently gets no workspace reserve and no attended-KV cap. Here it takes `chunked_workspace_profiled=False` and picks up both the `min(max_batch_size, max_num_tokens) * max_seq_len` cap and the reserve — at `max_batch_size=16` / `max_seq_len=262144` that is a sizeable slice of the KV budget on a config that works today.

If that's the deliberate fix for the full-gather path, could you note it and share a rough before/after block count for one Hopper config?

Required for this PR, so the capacity change is visible rather than a surprise.

Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/_util.py:1050]: Thanks for pointing this out. Adding a new reserve/cap on Hopper was outside the intended scope of this PR, so I restored its pre-PR policy in `32b7f72ea9`.

The new profiling requirement now applies only to SM100+. Dense MLA on Hopper with scheduler chunking enabled no longer gains an additional workspace reservation or attended-KV cap. Existing protection for non-chunked reuse and backend-declared unbounded workspace remains intact. SM90 is still excluded from chunk-aware profiling.

I added capacity regression coverage and updated the PR description. I don't have a measured Hopper before/after block count; this change removes the additional budget deduction you identified, while leaving Hopper's full-gather memory risk for a separate follow-up.

**Review by yizhang-nv — APPROVED** [review]

**Review by eopXD — COMMENTED** [review]
The profiling length is a **single** full chunk plus `max_num_tokens` plus a single KV block. This is assuming that the previous chunk's activations will not overlap with the follow-up chunk. [The current MLA code have the activation provided to the next attention call](https://github.com/NVIDIA/TensorRT-LLM/blob/c2997069caacfc1b97002093a40b5d4d69a1e4fd/tensorrt_llm/_torch/attention/mla.py#L1284). So overlapping happens.

I think we should adjust the profiling to be of two iterations, with a longer length of request, implying **two** full chunks.

**Comment by zhaoyangwang-nvidia** [comment]
Thanks for the review! I agree that profiling should cover the activation overlap between two full chunks. Updated in `ee2b2b5dbb`: the profiling request now builds a cached prefix of at least two full KV chunks, followed by a full `max_num_tokens` query, so both full chunks are processed in the same forward pass.

**Review by eopXD — APPROVED** [review]
Looks good. Thank you for the PR!

**Comment by eopXD** [comment]
Another thought to the profiling is that the estimation of things that scale with "batch" will not be covered.
Logits scale with batch:

```
                                            Single request    Many requests
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━
   Requests                                              1              128
  ───────────────────────────────────────  ────────────────  ───────────────
   Prompt tokens per request                         8,192               64
  ───────────────────────────────────────  ────────────────  ───────────────
   Total processed tokens                            8,192            8,192
  ───────────────────────────────────────  ────────────────  ───────────────
   Positions needing next-token sampling                 1              128
  ───────────────────────────────────────  ────────────────  ───────────────
   FP32 logits shape                           [1, 131072]    [128, 131072]
  ───────────────────────────────────────  ────────────────  ───────────────
   FP32 logits memory                              0.5 MiB           64 MiB
```

The PR currently limits batch size to 16. So I think a follow-up will be good to cover this case as well.
This is not a blocker for this current pull request though.

**Comment by zhaoyangwang-nvidia** [comment]
> Another thought to the profiling is that the estimation of things that scale with "batch" will not be covered. Logits scale with batch:
> 
> ```
>                                             Single request    Many requests
>   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━
>    Requests                                              1              128
>   ───────────────────────────────────────  ────────────────  ───────────────
>    Prompt tokens per request                         8,192               64
>   ───────────────────────────────────────  ────────────────  ───────────────
>    Total processed tokens                            8,192            8,192
>   ───────────────────────────────────────  ────────────────  ───────────────
>    Positions needing next-token sampling                 1              128
>   ───────────────────────────────────────  ────────────────  ───────────────
>    FP32 logits shape                           [1, 131072]    [128, 131072]
>   ───────────────────────────────────────  ────────────────  ───────────────
>    FP32 logits memory                              0.5 MiB           64 MiB
> ```
> 
> The PR currently limits batch size to 16. So I think a follow-up will be good to cover this case as well. This is not a blocker for this current pull request though.

@eopXD Thanks for pointing this out! I created [TRTLLM-16535](https://jirasw.nvidia.com/browse/TRTLLM-16535) to follow up on batch-dependent memory coverage in profiling.

I’m also reviewing our GPU buffer management and warmup process more broadly to reduce OOM risk, tracked under [TRTLLM-16463](https://jirasw.nvidia.com/browse/TRTLLM-16463). If you have suggestions or know of other potential OOM risks, please feel free to add tasks to the epic. I’ll work through them one by one.

**Comment by eopXD** [comment]
Thank you for spreading out the work! The roadmap looks good to me.

**Review by lori-ren — APPROVED** [review]

**Review by xinhe-nv — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]
